craft_parts/executor/collisions.py

```diff
@@ -97,11 +101,81 @@ def _get_candidate_from_install_dir(
         contents=part_contents,
         source_dir=part.part_install_dirs[partition],
         permissions=part.spec.permissions,
+        is_overlay=False,
     )
 
 
-def _check_for_stage_collisions_per_partition(
+def _get_overlay_layer_contents(source: pathlib.Path) -> tuple[set[str], set[str]]:
```

Comment:
I think @upils has implemented similar functionality in his migration work, maybe we can consolidate in the future.